server/stream.go

```diff
@@ -1043,15 +1043,30 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 		hdr = msg[:pc.pa.hdr]
 		msg = msg[pc.pa.hdr:]
 	}
-	seq, ts, err = store.StoreMsg(subject, hdr, msg)
-	if err == nil && seq > 0 {
-		mset.lseq = seq
-		mset.lmsgId = msgId
-	}
+
+	// Preemptively store the correct values here, will correct if seq guess was wrong.
+	olseq, olmsgId := mset.lseq, mset.lmsgId
+	mset.lseq++
+	mset.lmsgId = msgId
 
 	// We hold the lock to this point to make sure nothing gets between us since we check for pre-conditions.
 	mset.mu.Unlock()
 
+	// Can not hold lock here since storeUpdates() could need lock.
+	seq, ts, err = store.StoreMsg(subject, hdr, msg)
+
+	// Put back if we did not complete correctly.
+	if mset.lseq != seq {
```

Comment:
you can't access mset.lseq unprotected..

Also here are some sequences that worry me. Multiple producers (different server readloops) send to same stream:
- thread 1 does mset.lseq++, which say it then equal to 1, releases lock but does not call StoreMsg() yet.
- thread 2 does mset.lseq++, which means it is now equal to 2, releases and calls StoreMsg(), which returns seq==1.
- thread 1 does call StoreMsg(), which returns seq==2, grabs the lock and says (assuming that the check is now under lock and does not use the value that was evaluated before unlocking) mset.lseq == seq, so it will leave that alone.
- thread 2 resumes, get lock and check that mset.lseq(==2) is different from seq(==1), so it will then do mset.lseq=1, which is wrong!

If you capture mset.lseq value after mset.lseq++ and before releasing lock, then you have still a problem:
- thread 1 does mset.lseq++ => savedLseq==1, releases lock
- thread 2 does mset.lseq++ => savedLseq==2, releases lock
- thread 2 calls StoreMsg(), returns seq==1
- thread 1 calls StoreMsg(), returns seq==2
- thread 1 grabs lock see that savedLseq(==1) != seq(==2), so set mset.lseq=2
- thread 2 grabs lock, see that savedLseq(==2) != seq(==1), so set mset.lseq=1

Reply:
Good catch, let me think some more about it.

Reply:
ok hopefully third time is a charm.